scripts/performance/benchmark-s3-transfer

```diff
@@ -0,0 +1,243 @@
+        :param command_args: The arguments for the command.
+        :param transfer_type: The transfer type ('s3s3', 's3local', 'locals3')
+        """
+        self.result_dir = os.path.join(result_dir, command_name, transfer_type)
+        self.command_name = command_name
+        self.full_command = 'aws s3 %s %s' % (
+            command_name, command_args)
+        self.times_run = 0
+        self.transfer_type = transfer_type
+
+    def run_benchmark(self, benchmark_script, dry_run=False):
+        """Runs the command using the benchmark script.
+
+        :type benchmark_script: str
+        :param benchmark_script:
+
+        :type dry_run: bool
+        :param dry_run: If true, the command will be printed, but not run.
+        """
+        if not os.path.exists(self.result_dir):
+            os.makedirs(self.result_dir)
+        self.times_run += 1
+        out_file = 'performance%05d.csv' % self.times_run
+        out_file = os.path.join(self.result_dir, out_file)
+        benchmark_args = [
+            benchmark_script, self.full_command, '--output-file', out_file]
+        print(benchmark_args[1])
+
+        if not dry_run:
+            subprocess.check_call(benchmark_args)
+
+
+def _get_benchmark_script():
+    """Retrieves the benchmark script location from s3transfer if possible."""
+    s3transfer_directory = os.path.dirname(s3transfer.__file__)
+    s3transfer_directory = os.path.dirname(s3transfer_directory)
+    scripts_directory = 'scripts/performance'
+    scripts_directory = os.path.join(s3transfer_directory, scripts_directory)
+    benchmark_script = os.path.join(scripts_directory, 'benchmark')
+
+    if os.path.isfile(benchmark_script):
+        return benchmark_script
+    else:
+        return None
+
+
+def _get_cases(case_file, benchmark_script):
+    """Loads performance test cases from the case file."""
+    with open(case_file, 'r') as f:
+        case_definitions = json.load(f)
+
+    cases = []
+    for definition in case_definitions:
+        definition['benchmark_script'] = benchmark_script
+        cases.append(TransferPerformanceTestCase(**definition))
+
+    return cases
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        'cases', help='A json file defining the test cases to run.'
+    ),
+    parser.add_argument(
+        '-b', '--benchmark-script',
+        help=('The benchmark script to run the commands with. This should be '
+              'from s3transfer.'))
+    parser.add_argument(
+        '-o', '--result-dir', default='results',
+        help='The directory to output performance results to.'
+    ),
+    parser.add_argument(
+        '--dry-run', default=False, action='store_true',
+        help='If set, commands will only be printed out, not executed.'
+    )
+    args = parser.parse_args()
+
+    script = args.benchmark_script
+    if script is None:
+        script = _get_benchmark_script()
+
+    if script is None:
```

Comment:
You could just fold all this into the `required=True`, and a custom `type=your_validator` in the `add_argument` call for `--benchmark-script`.